python/pyspark/sql/connect/catalog.py

```diff
@@ -0,0 +1,46 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import NamedTuple, Optional, TYPE_CHECKING, List
+
+if TYPE_CHECKING:
+    from pyspark.sql.connect.session import SparkSession
+
+
+class Database(NamedTuple):
+    name: str
+    catalog: Optional[str]
+    description: Optional[str]
+    locationUri: str
+
+
+class Catalog:
+    """
+    User-facing catalog API, accessible through `SparkSession.catalog`.
+    """
+
+    def __init__(self, sparkSession: "SparkSession") -> None:
+        self._sparkSession = sparkSession
+
+    def listDatabases(self) -> List[Database]:
+        rows = self._sparkSession.sql("SHOW DATABASES").collect()
+        databases = []
+        for row in rows:
+            databases.append(
+                Database(name=row["namespace"], catalog=None, description=None, locationUri="")
```

Comment:
The first difference is there is `namespace` in the return row which should be `name`.

The second difference is there is no `catalog`, `description` and `localtionUrl` in the SQL returned row.

Reply:
Can we get them via `"DESCRIBE NAMESPACE %s" % ["namespace"]`? Seems like that's what we're doing in Scala API: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala#L81-L91

Reply:
The issue is that it would be pretty inefficient if there are many databases because we have to invoke `sql` multiple times. From a cursory look, there are many places like this (e.g., listing tables).

We could implement these fast as the first version without adding a lot of protobuf messages vs it would be pretty slow due to multiple sql invocations. e.g., 1000 tables, it takes around 7~8 secs in my local with pure Spark.

My suggestion is to quickly implement this, and switch it in the future but let me know if you guys have a different through @hvanhovell @grundprinzip

Reply:
yes I left a separate comment which says the same thing. Basically it would be N + 1 RPC to make this work where N is the number of databases.